lib/spack/spack/installer.py

```diff
@@ -0,0 +1,1533 @@
+            # assume using a separate (failed) prefix lock file.
+            if pkg_id in self.failed or spack.store.db.prefix_failed(spec):
+                tty.warn('{0} failed to install'.format(pkg_id))
+                self._update_failed(task)
+                continue
+
+            # Attempt to get a write lock.  If we can't get the lock then
+            # another process is likely (un)installing the spec or has
+            # determined the spec has already been installed (though the
+            # other process may be hung).
+            ltype, lock = self._ensure_locked('write', pkg)
+            if lock is None:
+                # Attempt to get a read lock instead.  If this fails then
+                # another process has a write lock so must be (un)installing
+                # the spec (or that process is hung).
+                ltype, lock = self._ensure_locked('read', pkg)
+
+            # Requeue the spec if we cannot get at least a read lock so we
+            # can check the status presumably established by another process
+            # -- failed, installed, or uninstalled -- on the next pass.
+            if lock is None:
+                self._requeue_task(task)
+                continue
+
+            # Determine state of installation artifacts and adjust accordingly.
+            self._check_install_artifacts(task, keep_prefix, keep_stage,
+                                          restage)
+
+            # Flag an already installed pkg
+            if pkg_id in self.installed:
+                # Downgrade to a read lock to preclude another processes
+                # from uninstalling the pkg until we're done.
+                #
+                # In the off chance we cannot get a read lock, then another
+                # process has probably taken a write lock between our releasing
+                # the write and acquiring the read.
+                ltype, lock = self._ensure_locked('read', pkg)
+                if lock is not None:
+                    self._update_installed(task)
+                    _print_installed_pkg(pkg.prefix)
+                else:
+                    # Since we cannot assess their intentions at this point,
+                    # requeue the task so we can re-check the status presumably
+                    # established by the other process -- failed, installed,
+                    # or uninstalled -- on the next pass.
+                    self._requeue_task(task)
+                continue
+
+            # Having a read lock on an uninstalled pkg may mean another
+            # process completed an uninstall of the software between the
+            # time we failed to acquire the write lock and the time we
+            # took the read lock.
+            #
+            # Requeue the task so we can check the status presumably
+            # established by the other process -- failed, installed, or
+            # uninstalled -- on the next pass.
```

Comment:
As mentioned above, I reworked this section and removed the offending words.

Not sure it's worth saying we want to avoid duplicating the write lock + install logic here given the reference to the next pass.